### ISSUE TITLE
Local node module (not npm registry) does not load with --node_modules-dir flag

### ISSUE THREAD

**Issue by abflow** [denoland/deno#18474]
I have a node module `test_node` which is locally installed within the deno project into a `node_modules` folder via:

`npm install --save ../test_node`

package.json:
```json
{
  "dependencies": {
    "test_node": "file:../test_node"
  }
}
```

When I run with the [node-modules-dir](https://deno.land/manual@v1.28.0/node/npm_specifiers#--node-modules-dir-flag) flag:
`deno run -A --node-modules-dir index.js`

I get the following error:
`error: npm package 'test_node' does not exist`

**Comment by abflow** [comment]
Ok, I had previously used into my `index.js` file:

`import Test from "npm:test_node";`

But now, when I simply used:
`import Test from "test_node";`

This time, I get the following error:
```
error: Parsing version constraints in the application-level package.json is more strict at the moment.
Not implemented scheme 'file'
```

**Comment by abflow** [comment]
One workaround could be to use the import map for loading a local node module without using `npm install`.

That would be possible if deno could somehow deduce internally from the import map value that it has to deal with a node module:
deno.json
```
{
	"imports": {
		"npm:test_node": "../test_node/index.js"
	}
}
```

Presently, this import map yields the following error:
`error: Uncaught SyntaxError: The requested module 'npm:test_node' does not provide an export named 'default'
`

**Comment by jsejcksn** [comment]
@abflow I didn't find anything indicating that this has been resolved, so I don't think it should be closed — AFAICT, Deno can't yet load Node modules from the local filesystem.

Ref: https://deno.land/manual@v1.32.3/node/package_json

<details>
<summary>Repro:</summary>

## Local Node module

`./local_tools/package.json`:
```json
{
  "name": "local_tools",
  "version": "0.1.0",
  "type": "module",
  "main": "index.mjs"
}

```

`./local_tools/index.mjs`:
```js
export { default as shout } from "./shout.mjs";

```

`./local_tools/shout.mjs`:
```js
/** @param {string} text */
function shout(text) {
  console.log(text.toUpperCase());
}

export default shout;

```


## Deno app

`./deno_app/package.json`:
```json
{
  "name": "deno_app",
  "version": "0.1.0",
  "dependencies": {
    "local_tools": "file:../local_tools"
  }
}

```

`./deno_app/main.ts`:
```ts
import { shout } from "local_tools";

shout("hello world");

```


## In the terminal

```
% pwd           
/Users/deno/example/deno_app

% deno --version
deno 1.32.3 (release, aarch64-apple-darwin)
v8 11.2.214.9
typescript 5.0.3

% deno run main.ts
error: Parsing version constraints in the application-level package.json is more strict at the moment.

Not implemented scheme 'file'
    at file:///Users/deno/example/deno_app/main.ts:1:23

```

</details>

**Comment by denizdogan** [comment]
I agree with @jsejcksn, and I don't think this is a minor issue, I'd say it's a pretty major issue, especially considering #18616. It seems that Deno can handle neither private npmjs packages nor local Node.js packages.

**Comment by abflow** [comment]
@jsejcksn @denizdogan You're right... sorry about that. I reopen the issue.

**Comment by festusyuma** [comment]
Hello, is there any update on this issue. Thank you

**Comment by dsherret** [comment]
No, there was a bit of work done on the patch feature for JSR (https://github.com/denoland/deno/pull/25068), but not for npm yet (I tried, but cut it out of the scope of the initial implementation).

**Comment by maheshbansod** [comment]
Are there any work-arounds? - this would be helpful for testing/debugging an NPM library with a deno project locally.

**Comment by SebastienGllmt** [comment]
I found a hacky workaround for this

1. Set `"nodeModulesDir": "auto",` in your root `deno.json` (this will get a `node_modules` folder generated)
2. Place your local dependency somewhere on your disk (say file path `./my-dep`)
3. Create a symlink from your folder to your `node_modules`

Here is an example of a deno task you can run to create the symblink assuming you placed the local dependency at the root of your deno project

```
"link:forge": "bash -c 'if ! [ -L ./node_modules/my-dep ]; then ln -s ../my-dep ./node_modules/my-dep; fi'",
```

**WARNING**: this may not always work. Notably, it has two downsides:
1. You will have to relink if you ever clear your `node_modules` (I don't think there is a good way to do this automatically for a local folder given deno's lack of non-npm postinstall script support)
2. This won't update your `deno.lock` (although presumably fewer things depend on this file being accurate)

**Comment by DreamwareDevelopment** [comment]
Trying to use Deno in Nx, total nightmare, thought this may help.

**Comment by SebastienGllmt** [comment]
Note: there has been progress on this via https://github.com/denoland/deno/pull/28396 which is part of the Deno 2.2.4 release

> In order to use these, you must not set --node-modules-dir=... when using a package.json and it should use the default of --node-modules-dir=manual.

**Comment by dsherret** [comment]
The original issue is now fixed. Let me know if it still occurs or please open new issues for other related issues.
